File: backend/analysis.py
"""Оценка сообщений моделью: очередь, повторы, фоновая переоценка.

Один рабочий поток на процесс — этого достаточно (docs/04-decisions.md, решение №2).
Сообщение никогда не пропадает из ленты из-за ошибки модели: после трёх неудач
оно закрывается как NORMAL с пометкой «Оценка недоступна»."""
import logging
import queue
import threading
import time

from sqlalchemy import select
from sqlalchemy.orm import Session

from backend import config
from backend.db import SessionLocal
from backend.events import bus
from backend.llm.provider import AnalysisRequest, LLMUnavailable, analyze
from backend.models import Connection, Message, OverrideLog, User, utcnow
from backend.serializers import message_out

log = logging.getLogger(__name__)

_queue: "queue.Queue[int]" = queue.Queue()
_worker: threading.Thread | None = None
_worker_lock = threading.Lock()


def enqueue(message_id: int) -> None:
    """Поставить сообщение в очередь на оценку."""
    _queue.put(message_id)
    ensure_worker()


def ensure_worker() -> None:
    global _worker
    with _worker_lock:
        if _worker is None or not _worker.is_alive():
            _worker = threading.Thread(target=_run, name="analysis-worker", daemon=True)
            _worker.start()


def _run() -> None:  # pragma: no cover - цикл потока, логика в process_message
    while True:
        message_id = _queue.get()
        try:
            with SessionLocal() as db:
                process_message(db, message_id)
        except Exception:
            log.exception("Не удалось оценить сообщение %s", message_id)
        finally:
            _queue.task_done()


def recent_overrides(db: Session, user: User, limit: int | None = None) -> list[tuple[str, str]]:
    """Последние ручные исправления пользователя — обратная связь для модели."""
    limit = limit or config.LLM_OVERRIDE_HISTORY
    rows = db.execute(
        select(Message.subject, OverrideLog.to_level)
        .join(Message, Message.id == OverrideLog.message_id)
        .join(Connection, Connection.id == Message.connection_id)
        .where(Connection.user_id == user.id)
        .order_by(OverrideLog.created_at.desc(), OverrideLog.id.desc())
        .limit(limit)
    ).all()
    return [(subject, level) for subject, level in rows]


def build_request(db: Session, message: Message) -> AnalysisRequest:
    user = message.connection.user
    sender = f"{message.sender_name} <{message.sender_addr}>".strip()
    return AnalysisRequest(
        criteria=user.criteria or "",
        sender=sender,
        subject=message.subject,
        body=message.body,
        source=message.connection.kind,
        overrides=recent_overrides(db, user),
    )


def process_message(db: Session, message_id: int, sleep=time.sleep) -> Message | None:
    """Оценить одно сообщение. Три повтора с задержками 2с / 8с / 30с."""
    message = db.get(Message, message_id)
    if message is None:
        return None

    request = build_request(db, message)
    result = None
    last_error: Exception | None = None
    for attempt, delay in enumerate(config.LLM_RETRY_DELAYS):
        try:
            result = analyze(request)
            break
        except LLMUnavailable as error:
            last_error = error
            log.warning("Модель не ответила (попытка %s): %s", attempt + 1, error)
            if attempt < len(config.LLM_RETRY_DELAYS) - 1:
                sleep(delay)

    if result is None:
        log.error("Оценка недоступна для сообщения %s: %s", message_id, last_error)
        # У сообщения, которое уже оценивалось, прежнюю оценку не стираем:
        # при переоценке потерять готовую карточку хуже, чем показать старую.
        if message.analyzed_at is None:
            message.level = "NORMAL"
            message.category = ""
            message.deadline_text = ""
            message.summary = ""
            message.needs_reply = False
            message.needs_action = False
        message.status = "DONE"
        message.analysis_failed = True
    else:
        message.status = "DONE"
        message.level = result.level
        message.category = result.category
        message.deadline_text = result.deadline
        message.needs_reply = result.needs_reply
        message.needs_action = result.needs_action
        message.summary = result.summary
        message.analysis_failed = False
    message.analyzed_at = utcnow()
    db.commit()
    db.refresh(message)
    bus.publish(
        message.connection.user_id,
        "message.analyzed",
        message_out(message).model_dump(mode="json"),
    )
    return message


def queue_reanalysis(db: Session, user: User) -> int:
    """Смена критериев: все сообщения пользователя уходят на переоценку.
    Ручные исправления (level_override) при этом не трогаем — решение №15."""
    messages = db.execute(
        select(Message)
        .join(Connection)
        .where(Connection.user_id == user.id)
        .order_by(Message.received_at.desc())
    ).scalars().all()
    for message in messages:
        message.status = "PROCESSING"
    db.commit()
    for message in messages:
        enqueue(message.id)
    return len(messages)

File: backend/internal/services/analysis/analysis.go
// Package analysis — оценка сообщений моделью: очередь, повторы, фоновая
// переоценка. Один рабочий поток на процесс, этого достаточно
// (docs/04-decisions.md, решение №2).
//
// Сообщение никогда не пропадает из ленты из-за ошибки модели: после трёх
// неудач оно закрывается как NORMAL с пометкой «Оценка недоступна».
package analysis

import (
	"errors"
	"log"
	"personalinbox/internal/exceptions"
	"sync"
	"time"

	"personalinbox/internal/core"
	"personalinbox/internal/events"
	"personalinbox/internal/openai"
	"personalinbox/internal/postgres"
	"personalinbox/internal/schemas"
)

// Enqueuer — то, куда ручки складывают сообщения на оценку.
// В тестах подставляется список, а не рабочий поток.
type Enqueuer interface {
	Enqueue(messageID int64)
}

// queue — неограниченная очередь id: переоценка ставит в неё сразу все
// сообщения пользователя, и терять их из-за размера буфера нельзя.
type queue struct {
	mu     sync.Mutex
	cond   *sync.Cond
	items  []int64
	closed bool
}

func newQueue() *queue {
	q := &queue{}
	q.cond = sync.NewCond(&q.mu)
	return q
}

func (q *queue) push(id int64) {
	q.mu.Lock()
	defer q.mu.Unlock()
	q.items = append(q.items, id)
	q.cond.Signal()
}

// pop ждёт следующий id. Второе значение — false, если очередь закрыта.
func (q *queue) pop() (int64, bool) {
	q.mu.Lock()
	defer q.mu.Unlock()
	for len(q.items) == 0 && !q.closed {
		q.cond.Wait()
	}
	if len(q.items) == 0 {
		return 0, false
	}
	id := q.items[0]
	q.items = q.items[1:]
	return id, true
}

func (q *queue) close() {
	q.mu.Lock()
	defer q.mu.Unlock()
	q.closed = true
	q.cond.Broadcast()
}

// Worker — рабочий поток оценки.
type Worker struct {
	db       *postgres.DB
	bus      *events.Bus
	analyzer openai.Analyzer

	// Sleep и RetryDelays вынесены наружу, чтобы тесты не ждали 40 секунд.
	Sleep       func(time.Duration)
	RetryDelays []time.Duration

	queue   *queue
	once    sync.Once
	done    chan struct{}
	stopped sync.Once
}

// NewWorker собирает воркер поверх базы, шины событий и адаптера модели.
func NewWorker(db *postgres.DB, bus *events.Bus, analyzer openai.Analyzer) *Worker {
	return &Worker{
		db:          db,
		bus:         bus,
		analyzer:    analyzer,
		Sleep:       time.Sleep,
		RetryDelays: core.LLMRetryDelays,
		queue:       newQueue(),
		done:        make(chan struct{}),
	}
}

// Enqueue ставит сообщение в очередь на оценку и поднимает поток, если он спит.
func (w *Worker) Enqueue(messageID int64) {
	w.queue.push(messageID)
	w.Start()
}

// Start запускает рабочий поток. Повторный вызов ничего не делает.
func (w *Worker) Start() {
	w.once.Do(func() {
		go w.run()
	})
}

// Stop останавливает поток: нужен корректному завершению сервера и тестам.
func (w *Worker) Stop() {
	w.stopped.Do(func() {
		w.queue.close()
	})
}

func (w *Worker) run() {
	defer close(w.done)
	for {
		id, ok := w.queue.pop()
		if !ok {
			return
		}
		if _, err := w.ProcessMessage(id); err != nil {
			log.Printf("не удалось оценить сообщение %d: %v", id, err)
		}
	}
}

// BuildRequest собирает то, что уходит в модель об одном сообщении.
func (w *Worker) BuildRequest(message *postgres.Message) (openai.Request, error) {
	connection, err := w.db.ConnectionByID(message.ConnectionID)
	if err != nil {
		return openai.Request{}, err
	}
	user, err := w.db.UserByID(connection.UserID)
	if err != nil {
		return openai.Request{}, err
	}
	overrides, err := w.db.RecentOverrides(user.ID, core.LLMOverrideHistory)
	if err != nil {
		return openai.Request{}, err
	}
	sender := message.SenderName + " <" + message.SenderAddr + ">"
	return openai.Request{
		Criteria:  user.Criteria,
		Sender:    sender,
		Subject:   message.Subject,
		Body:      message.Body,
		Source:    connection.Kind,
		Overrides: overrides,
	}, nil
}

// ProcessMessage оценивает одно сообщение. Три повтора с задержками 2с / 8с / 30с.
// Отсутствующее сообщение — не ошибка: его могли удалить, пока оно ждало очереди.
func (w *Worker) ProcessMessage(messageID int64) (*postgres.Message, error) {
	message, err := w.db.MessageByID(messageID)
	if errors.Is(err, exceptions.ErrNotFound) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	request, err := w.BuildRequest(message)
	if err != nil {
		return nil, err
	}

	var result openai.Result
	var lastErr error
	succeeded := false
	for attempt, delay := range w.RetryDelays {
		result, lastErr = w.analyzer.Analyze(request)
		if lastErr == nil {
			succeeded = true
			break
		}
		log.Printf("модель не ответила (попытка %d): %v", attempt+1, lastErr)
		if attempt < len(w.RetryDelays)-1 {
			w.Sleep(delay)
		}
	}

	if !succeeded {
		log.Printf("оценка недоступна для сообщения %d: %v", messageID, lastErr)
		// У сообщения, которое уже оценивалось, прежнюю оценку не стираем:
		// при переоценке потерять готовую карточку хуже, чем показать старую.
		if message.AnalyzedAt == nil {
			message.Level = "NORMAL"
			message.Category = ""
			message.DeadlineText = ""
			message.Summary = ""
			message.NeedsReply = false
			message.NeedsAction = false
		}
		message.Status = "DONE"
		message.AnalysisFailed = true
	} else {
		message.Status = "DONE"
		message.Level = result.Level
		message.Category = result.Category
		message.DeadlineText = result.Deadline
		message.NeedsReply = result.NeedsReply
		message.NeedsAction = result.NeedsAction
		message.Summary = result.Summary
		message.AnalysisFailed = false
	}
	now := postgres.UTCNow()
	message.AnalyzedAt = &now
	if err := w.db.SaveMessage(message); err != nil {
		return nil, err
	}

	connection, err := w.db.ConnectionByID(message.ConnectionID)
	if err != nil {
		return nil, err
	}
	w.bus.Publish(connection.UserID, "message.analyzed", schemas.MessageOut(message))
	return message, nil
}

// QueueReanalysis — смена критериев: все сообщения пользователя уходят
// на переоценку. Ручные исправления (level_override) при этом не трогаем —
// решение №15.
func QueueReanalysis(db *postgres.DB, enqueuer Enqueuer, userID int64) (int, error) {
	ids, err := db.MarkAllProcessing(userID)
	if err != nil {
		return 0, err
	}
	for _, id := range ids {
		enqueuer.Enqueue(id)
	}
	return len(ids), nil
}
